File: password-vault/Models/PasswordEntry.m
#import "PasswordEntry.h"

NS_ASSUME_NONNULL_BEGIN

@implementation PasswordEntry

- (instancetype)initWithTitle:(NSString *)title
                     username:(NSString *)username
                     password:(NSString *)password
                        notes:(nullable NSString *)notes {
    if (self = [super init]) {
        _identifier = [NSUUID UUID];
        _title = title;
        _username = username;
        _password = password;
        _notes = notes;
    }
    return self;
}

- (instancetype)initWithIdentifier:(NSUUID *)identifier
                             title:(NSString *)title
                          username:(NSString *)username
                          password:(NSString *)password
                             notes:(nullable NSString *)notes {
    if (self = [super init]) {
        _identifier = identifier;
        _title = title;
        _username = username;
        _password = password;
        _notes = notes;
    }
    return self;
}

+ (BOOL)supportsSecureCoding {
    return YES;
}

- (void)encodeWithCoder:(NSCoder *)coder {
    [coder encodeObject:self.identifier forKey:@"id"];
    [coder encodeObject:self.title forKey:@"title"];
    [coder encodeObject:self.username forKey:@"username"];
    [coder encodeObject:self.password forKey:@"password"];
    [coder encodeObject:self.notes forKey:@"notes"];
}

- (nullable instancetype)initWithCoder:(NSCoder *)coder {
    if (self = [super init]) {
        _identifier = [coder decodeObjectOfClass:[NSUUID class] forKey:@"id"];
        _title = [coder decodeObjectOfClass:[NSString class] forKey:@"title"];
        _username = [coder decodeObjectOfClass:[NSString class]
                                        forKey:@"username"];
        _password = [coder decodeObjectOfClass:[NSString class]
                                        forKey:@"password"];
        _notes = [coder decodeObjectOfClass:[NSString class] forKey:@"notes"];
    }
    return self;
}

@end

NS_ASSUME_NONNULL_END

File: decryptor/PasswordEntry.swift
import Foundation

@objc class PasswordEntry: NSObject, NSSecureCoding {
    static var supportsSecureCoding: Bool { return true }

    let identifier: UUID
    var title: String
    var username: String
    var password: String
    var notes: String?

    init(title: String, username: String, password: String, notes: String?) {
        self.identifier = UUID()
        self.title = title
        self.username = username
        self.password = password
        self.notes = notes
        super.init()
    }

    required init?(coder: NSCoder) {
        guard let identifier = coder.decodeObject(of: NSUUID.self, forKey: "id") as? UUID,
            let title = coder.decodeObject(of: NSString.self, forKey: "title") as? String,
            let username = coder.decodeObject(of: NSString.self, forKey: "username") as? String,
            let password = coder.decodeObject(of: NSString.self, forKey: "password") as? String
        else {
            return nil
        }

        self.identifier = identifier
        self.title = title
        self.username = username
        self.password = password
        self.notes = coder.decodeObject(of: NSString.self, forKey: "notes") as? String
        super.init()
    }

    func encode(with coder: NSCoder) {
        coder.encode(identifier as NSUUID, forKey: "id")
        coder.encode(title as NSString, forKey: "title")
        coder.encode(username as NSString, forKey: "username")
        coder.encode(password as NSString, forKey: "password")
        if let notes = notes {
            coder.encode(notes as NSString, forKey: "notes")
        }
    }

    override var description: String {
        return """
            PasswordEntry:
              ID: \(identifier)
              Title: \(title)
              Username: \(username)
              Password: \(password)
              Notes: \(notes ?? "none")
            """
    }
}
